Add type to shader material constructor

stefnotch opened the pull request:
Fix https://forum.babylonjs.com/t/shadermaterial-constructor-type-safety/49030

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

stefnotch commented on packages/dev/core/src/Engines/thinEngine.ts:
```diff
@@ -2935,8 +2936,26 @@ export class ThinEngine {
+                ? baseName.fragment
+                : "fragmentToken" in baseName
+                ? baseName.fragmentToken
+                : baseName.fragmentSource || baseName;
```
If someone knows of a shorter way of making Typescript happy, we could do that.

Also, this technically changes the behaviour. Previously `{ vertexElement: "", vertex: "custom" }` would have caused the code to read "custom". Now it reads `""`.

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

stefnotch commented:
One CI failure is from https://github.com/BabylonJS/Babylon.js/blob/c05e7f07cddb0a1b6f86f6bb06cf8a34f0863b6a/packages/dev/core/src/Materials/shadowDepthWrapper.ts#L302-L307 . A property named `vertexToken` is being used there, and I have no idea if that's intentional or if it's a bug.
The property was undocumented.

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

Popov72 commented:
> A property named vertexToken is being used there, and I have no idea if that's intentional or if it's a bug.
The property was undocumented.

Yes, this is intentional, it's a way of uniquely identifying a vertex+fragment shader regardless of its defines list (normally, a vertex+fragment shader pair is uniquely defined by the defines list it uses). I think you can document it as `@internal` at this point, users probably won't need it.

Popov72 commented on packages/dev/core/src/Engines/thinEngine.ts:
```diff
@@ -2935,8 +2936,8 @@ export class ThinEngine {
     ): Effect {
-        const vertex = baseName.vertexElement || baseName.vertex || baseName.vertexToken || baseName.vertexSource || baseName;
-        const fragment = baseName.fragmentElement || baseName.fragment || baseName.fragmentToken || baseName.fragmentSource || baseName;
+        const vertex = typeof baseName === "string" ? baseName : baseName.vertexSource || baseName.vertexElement || baseName.vertex || (baseName as any).vertexToken;
```
You should keep the same order than previously: `baseName.vertexElement || baseName.vertex || baseName.vertexToken || baseName.vertexSource`.

Same thing for the `fragment` case.

stefnotch replied:
I intentionally changed this, so that it matches with the order in `Materials/effect.ts`. Is that alright?
(I also politely added that priority info to the IShaderPath docs.)

Popov72 replied:
Ok for me!

The "tokenXXX" check should come first.

stefnotch replied:
Fixed :)

Popov72 commented on packages/dev/core/src/Materials/shaderMaterial.ts:
```diff
@@ -35,6 +35,45 @@ import {
+ *  * object: `{ vertex: "custom", fragment: "custom" }`, used with `Effect.ShadersStore["customVertexShader"]` and `Effect.ShadersStore["customFragmentShader"]`
+ *  * string: `"./COMMON_NAME"`, used with external files COMMON_NAME.vertex.fx and COMMON_NAME.fragment.fx in index.html folder.
+ */
+export type IShaderPath =
```
You should move this definition to `effect.ts` instead, so that `thinEngine` and `WebGPUEngine` don't get a dependency on `ShaderMaterial`.

Popov72 commented on packages/dev/core/src/Materials/shaderMaterial.ts:
```diff
@@ -35,6 +35,45 @@ import {
+           * Used with shader code in script tags
+           */
+          fragmentElement?: string;
+      }
```
See my comment in the thread: add `vertexToken?: string` in the object, but document is as `/* @internal */` so that it's not visible to users.

Popov72 reviewed (changes requested):
Good job!

You will also have to update the `WebGPUEngine.createEffect` method.

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

bjsplat commented:
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

bjsplat commented:
Snapshot stored with reference name:
refs/pull/14908/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14908/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14908/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14908/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14908/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14908/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14908/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14908/merge#BCU1XR#0

bjsplat commented:
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14908/merge/testResults/webgpuplaywright/index.html

bjsplat commented:
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14908/merge/testResults/webgl2playwright/index.html

ryantrem commented:
> > A property named vertexToken is being used there, and I have no idea if that's intentional or if it's a bug.
> > The property was undocumented.
> 
> Yes, this is intentional, it's a way of uniquely identifying a vertex+fragment shader regardless of its defines list (normally, a vertex+fragment shader pair is uniquely defined by the defines list it uses). I think you can document it as `@internal` at this point, users probably won't need it.

@Popov72 if vertexToken/fragmentToken are intended to be internal only, would it make sense to leave them out of `IShaderPath` such that they are not expected to be passed into the `Effect` constructor, and have the `baseName` parameter of the `createEffect` engine functions be `IShaderPath & { vertexToken?: string; fragmentToken?: string}`? This would be effectively assuming Engine.createEffect is intended to be for internal use, and the Effect constructor is intended to be the public API.

Popov72 commented:
Good idea @ryantrem!

stefnotch commented:
Good idea, I changed it to `IShaderPath & { vertexToken?: string; fragmentToken?: string}`.

While I was at it, I noticed that the same sort of type could be added to the compute shaders. So I introduced a `IComputeShaderPath` and did all the changes there as well.

bjsplat commented:
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14908/merge/testResults/webgpuplaywright/index.html
